core/src/serve_repair.rs

```diff
@@ -432,19 +454,37 @@ impl ServeRepair {
         packet_batch: PacketBatch,
         response_sender: &PacketBatchSender,
         stats: &mut ServeRepairStats,
+        data_budget: &DataBudget,
     ) {
         // iter over the packets
-        packet_batch.iter().for_each(|packet| {
+        let mut last_processed_packet_index = None;
+        for (i, packet) in packet_batch.iter().enumerate() {
+            last_processed_packet_index = Some(i);
             if let Ok(request) = packet.deserialize_slice(..) {
                 stats.processed += 1;
                 let from_addr = packet.meta.socket_addr();
                 let rsp = Self::handle_repair(me, recycler, &from_addr, blockstore, request, stats);
-                stats.total_response_packets += rsp.as_ref().map(PacketBatch::len).unwrap_or(0);
+                let num_response_packets = rsp.as_ref().map(PacketBatch::len).unwrap_or(0);
                 if let Some(rsp) = rsp {
-                    let _ignore_disconnect = response_sender.send(rsp);
+                    let num_response_bytes = rsp.iter().map(|p| p.meta.size).sum();
+                    if data_budget.take(num_response_bytes) {
+                        stats.total_response_bytes += num_response_bytes;
+                        stats.total_response_packets += num_response_packets;
+                        let _ignore_disconnect = response_sender.send(rsp);
+                    } else {
+                        stats.dropped_requests += 1;
```

Comment:
Can we remove `last_processed_packet_index` and instead here update `stats.dropped_requests` correctly based on `i` and `packet_batch.len()`? i.e.
```rust
stats.dropped_requests += packet_batch.len() - i;
```